crypto/sm2/sm2_pmeth.c

```diff
@@ -209,9 +247,32 @@ static int pkey_sm2_ctrl_str(EVP_PKEY_CTX *ctx,
     return -2;
 }
 
+static int pkey_sm2_digest_custom(EVP_PKEY_CTX *ctx, EVP_MD_CTX *mctx)
+{
+    uint8_t z[EVP_MAX_MD_SIZE];
+    SM2_PKEY_CTX *smctx = ctx->data;
+    EC_KEY *ec = ctx->pkey->pkey.ec;
+    const EVP_MD *md = EVP_MD_CTX_md(mctx);
+
+    if (smctx->id_len == -1) {
+        /*
+         * If no ID is set, reject to compute the 'Z' value. This allows to
+         * set a null-ID to SM2, although which is not clear in the relevant
+         * specifications. We implement as this to stay flexibility.
```

Comment:
Perhaps reword as: "An ID value must be set. The specifications are not clear whether a NULL is allowed. We only allow it if set explicitly for maximum flexibility"